libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,42 +112,121 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
+                        if (oAuthTokenTask.Status == TaskStatus.RanToCompletion && oAuthTokenTask.Result.access_token == oAuthToken.access_token)
                         {
-                            return oAuthToken.access_token;
+                            oAuthTokenTask = _getCurrentTokenTask(forceRefresh: true);
                         }
                     }
+
+                    // always check the autorefresh
+                    CheckAutoRefreshToken();
+                }
+            }
+
+            // if we have an oAuthTokenTask then we need to await it
+            if (oAuthTokenTask != null)
+            {
+                oAuthToken = await oAuthTokenTask;
+                tokenCache[TokenCacheKey] = oAuthToken;
+            }
+
+            return oAuthToken?.access_token;
+        }
+
+        private void CheckAutoRefreshToken()
+        {
+            // get the current autorefreshTime for this key
+            DateTime refreshTime;
+            if (autoRefreshTimes.TryGetValue(TokenCacheKey, out refreshTime))
+            {
+                // if we are past the refresh time
+                if (DateTime.UtcNow > refreshTime)
+                {
+                    // set new refresh time (this keeps only one outstanding refresh Task at a time)
+                    autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
+
+                    // background task to refresh the token
+                    // NOTE: This is not awaited, but is observed with the ContinueWith() clause.  
+                    // It is gated by the AutoRefreshTimes[] array
+                    RefreshTokenAsync()
+                        .ContinueWith(task =>
+                        {
+                            // observe the background task and put in cache when done
+                            if (task.Status == TaskStatus.RanToCompletion)
+                            {
+                                // update the cache with completed task so all new requests will get it
+                                tokenTaskCache[TokenCacheKey] = task;
+                            }
+                            else
+                            {
+                                // it failed, shorten the refresh time for another task to try again in a 30s
+                                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + TimeSpan.FromSeconds(30);
+                            }
+                        });
                 }
             }
+        }
 
-            // We need to refresh the token, because:
-            // 1. The user requested it via the forceRefresh parameter
-            // 2. We have it, but it's expired
-            // 3. We don't have it in the cache. 
+        /// <summary>
+        /// Do not call this except from GetTokenRefreshAsync() 
+        /// </summary>
+        /// <returns></returns>
+        private Task<OAuthResponse> _getCurrentTokenTask(bool forceRefresh)
```

Comment:
Using an underscore for private methods is not a typical convention in .NET land. Suggest just `GetCurrentTokenTask`.


Also, your comment of `"Do not call this except from GetTokenRefreshAsync"` is literally why local functions were added to the language. I would _highly_ suggest this be made into a local function.

Reply:
hmm...ok, that's a good idea I forget that feature exists.  I still code to .net 2.0 LOL 

---
In reply to: [198883214](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198883214) [](ancestors = 198883214)

Reply:
Actually, I hate the local function feature.  When you create a local function it is a leaky abstraction in that the local function picks up the scope of the variables of the function it is defined in.  IMHO this breaks the isolation which is part of what makes a method useful.  If I need common data it should be accessed from the object, not the scope.  Point in fact, the variable name forceRefresh is in conflict with the calling context, but my whole usage pattern is that I change the forceRefresh policy depending on my state. 

I will not be moving this.

---
In reply to: [198906624](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198906624) [](ancestors = 198906624,198883214)